development/ts-migration-dashboard/scripts/build-module-partitions.ts

```diff
@@ -0,0 +1,336 @@
+  entryModuleIds: string[],
+  allowedFilePaths: string[],
+): Record<string, Module> {
+  // Our overall goal is that we want to partition the graph into different
+  // sections. We want to find the leaves of the graph — that is, files that
+  // depend on no other files — then the dependents of the leaves — those files
+  // that depend on the leaves — then the dependents of the dependents, etc. We
+  // can derive this information by traversing the graph, and for each module we
+  // encounter, recording the number of modules it takes to reach that module.
+  // We call this number the **level**.
+  //
+  // We will discuss a couple of optimizations we've made to ensure that graph
+  // traversal is performant.
+  //
+  // Naively, in order to calculate the level of each module, we need to follow
+  // that module's dependencies, then the dependencies of the dependencies,
+  // etc., until we hit leaves. Essentially, we need to follow each connection
+  // in the graph. However, this creates a performance problem, because in a
+  // large system a file could get imported multiple ways (this is the nature of
+  // a graph: each node can have multiple incoming connections and multiple
+  // outgoing connections). For instance:
+  //
+  // - `foo.js` could import `bar.js` which could import `baz.js`
+  // - `foo.js` could also import `baz.js` directly
+  // - `foo.js` could also import `bar.js` which imports `qux.js` which imports `baz.js`
+  //
+  // In this case, even if there are few modules in a system, a subset of those
+  // modules may be visited multiple times in the course of traversing
+  // connections between all of them. This is costly and unnecessary.
+  //
+  // To address this, as we are traversing the graph, we record modules we've
+  // visited along with the level when we visited it. If we encounter a module
+  // again through some other path, and the level this time is less than the
+  // level we've already recorded, then we know we don't need to revisit that
+  // module or — crucially — any of its dependencies. Therefore we can skip
+  // whole sections of the graph.
+  //
+  // In addition, in a large enough system, some files could import files that end
+  // up importing themselves later on:
+  //
+  // - `foo.js` could import `bar.js`, which imports `baz.js`, which imports `foo.js`, which...
+  //
+  // These are called circular dependencies, and we detect them by tracking the
+  // files that depend on a file (dependents) in addition to the files on which
+  // that file depends (dependencies). In the example above, `baz.js` has a
+  // dependency `foo.js`, and its chain of dependents is `bar.js` -> `foo.js`
+  // (working backward from `baz.js`). The chain of dependents of `baz.js`
+  // includes `foo.js`, so we say `foo.js` is a circular dependency of `baz.js`,
+  // and we don't need to follow it.
+
+  const modulesToFill: Module[] = entryModuleIds.map((moduleId) => {
```

Comment:
I regret to say, upon looking back at this, that there are no tests for this file (for anything, but especially this). I don't know that I have time to add tests right now for this but I'm happy to explain what this is doing over a call.